src/reflect/scala/reflect/internal/util/ReusableInstance.scala

```diff
@@ -10,28 +10,38 @@
  * additional information regarding copyright ownership.
  */
 
-package scala
-package reflect
-package internal
-package util
+package scala.reflect.internal.util
 
-/** A wrapper for a re-entrant, cached instance of a value of type `T`.
+import scala.collection.mutable.ArrayBuffer
+import scala.util.chaining._
+
+/** A wrapper for a list of cached instances of a type `T`.
+  * The wrapper is recursion-reentrant: several instances are kept, so
+  * at each depth of reentrance we are reusing the instance for that.
+  *
+  * An instance is created upon creating this object, and more instances
+  * are allocated dynamically, on demand, when reentrance occurs.
   *
   * Not thread safe.
   */
-final class ReusableInstance[T <: AnyRef](make: () => T, enabled: Boolean) {
-  private val cached = make()
-  private var taken = false
+final class ReusableInstance[T <: AnyRef] private (make: => T, enabled: Boolean) {
+  private[this] val cache = if (enabled) new ArrayBuffer[T](ReusableInstance.InitialSize).tap(_.addOne(make)) else null
+  private[this] var taken = 0
```

Comment:
I was going to complain that this would stymie inlining, but lo:
```scala
// make not private symbol accessed from inner classes, as well as
// symbols accessed from @inline methods
//
// See scala/bug#6552 for an example of why `sym.owner.enclMethod hasAnnotation ScalaInlineClass`
// is not suitable; if we make a method-local class non-private, it mangles outer pointer names.
def enclMethodIsInline = closestEnclMethod(currentOwner) hasAnnotation ScalaInlineClass
// scala/bug#8710 The extension method condition reflects our knowledge that a call to `new Meter(12).privateMethod`
//         with later be rewritten (in erasure) to `Meter.privateMethod$extension(12)`.
if ((currentClass != sym.owner || enclMethodIsInline) && !sym.isMethodWithExtension)
  sym.makeNotPrivate(sym.owner)
```